### PR TITLE
Add module: goat/taxonsearch

### LINKED ISSUE #1864: new module: goat/taxonsearch
### Is there an existing module for this?

- [X] I have searched for the existing module

### Is there an open PR for this?

- [X] I have searched for existing PRs

### Is there an open issue for this?

- [X] I have searched for existing issues

### Are you going to work on this?

- [X] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by alxndrdiaz**
## PR checklist

Closes #1864 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/goat/taxonsearch/main.nf [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    tuple val(meta), path("*.tsv"), emit: taxonsearch
+    path "versions.yml"           , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    input = taxa_file.name != 'NO_FILE' ? "-f ${taxa_file}" : "-t ${taxon}"
```
Comment (nvnieuwk) [inline modules/goat/taxonsearch/main.nf:23]: If I understand correctly, you are trying to determine whether a file is entered or not? If that's the case you can add an empty channel using `[]` as input and chance your check to `input = taxa_file.name ? "-f ${taxa_file}" : "-t ${taxon}"`. This is the default way on how to specify an optional channel :)

Comment (nvnieuwk) [reply modules/goat/taxonsearch/main.nf:23]: Same for the next two lines

File: tests/modules/goat/taxonsearch/main.nf [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+nextflow.enable.dsl = 2
+
+include { GOAT_TAXONSEARCH } from '../../../../modules/goat/taxonsearch/main.nf'
+
+//
+// Test with genus name (Canis)
+//
+workflow test_goat_taxonsearch_genus_name {
+
+    input_file = file('NO_FILE')
```
Comment (nvnieuwk) [inline tests/modules/goat/taxonsearch/main.nf:12]: So, change this to `[]` instead of `file('NO_FILE')`

**Review by nvnieuwk — CHANGES_REQUESTED** [review]

File: modules/goat/taxonsearch/main.nf [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    if (!taxon && taxa_file.name == 'NO_FILE') error "No input. Valid input: single taxon identifier or a .txt file with identifiers"
+    if (taxon && taxa_file.name != 'NO_FILE') error "Only one input is required: a single taxon identifier or a .txt file with identifiers"
+    """
+    goat-cli taxon search \\
+        $args \\
+        $input > ${prefix}.tsv
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        goat: \$(echo \$( goat-cli --version | cut -d' ' -f2 | head -1 2>&1) | sed 's/^.*goat //; s/Using.*\$//' ))
```
Comment (muffato) [inline modules/goat/taxonsearch/main.nf:33]:
```suggestion
        goat: \$(goat-cli --version | cut -d' ' -f2)
```

File: tests/modules/goat/taxonsearch/main.nf [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+      taxon = 'Canis',
+      []
+  ]
+    GOAT_TAXONSEARCH ( input )
+}
+
+//
+// Test with genus (Drosophila, fruit flies) using NCBI taxonomy ID
+//
+workflow test_goat_taxonsearch_genus_id {
```
Comment (nvnieuwk) [inline tests/modules/goat/taxonsearch/main.nf:23]: Could you also add a test with a taxa file?

**Review by nvnieuwk — CHANGES_REQUESTED** [review]

File: tests/modules/goat/taxonsearch/main.nf (hunk 2 above)
Comment (alxndrdiaz) [reply tests/modules/goat/taxonsearch/main.nf]: Done.

File: tests/modules/goat/taxonsearch/main.nf (hunk 4 above)
Comment (alxndrdiaz) [reply tests/modules/goat/taxonsearch/main.nf:23]: I guess is possible, I would need to create a test file in the [nf-core test datasets repo](https://github.com/nf-core/test-datasets) and then fetch the file from there, sorry for the late response.

Comment (nvnieuwk) [reply tests/modules/goat/taxonsearch/main.nf:23]: No problem! Are these taxa files big? Maybe post in the `test-data` slack channel before adding it. And see if it's okay to add :)

Comment (alxndrdiaz) [reply tests/modules/goat/taxonsearch/main.nf:23]: They only contain the NCBI tax IDs or binomial names to search (one per line,  limited to 500 lines max.), so it should be small.

Comment (nvnieuwk) [reply tests/modules/goat/taxonsearch/main.nf:23]: That should be fine then, I think

Comment (alxndrdiaz) [reply tests/modules/goat/taxonsearch/main.nf:23]: Hi, I just added the .txt file that will be used for the test, here is the PR: [test file for goat/taxonsearch module #590](https://github.com/nf-core/test-datasets/pull/590), please check it before merging it, I will add the test to the module today.

Comment (alxndrdiaz) [reply tests/modules/goat/taxonsearch/main.nf:23]: Done, I just added a tests using an input taxa file.

Comment (alxndrdiaz) [reply tests/modules/goat/taxonsearch/main.nf:23]: > Could you also add a test with a taxa file?

Done, I just added a test using an input taxa file

**Review by nvnieuwk — APPROVED** [review]
LGTM!!!

**Review by muffato — APPROVED** [review]

**Pull request merged**